### DISCUSSION TITLE
Problem flashing Tasmota to ESP32

### DISCUSSION THREAD

**Discussion by laurentppol** [arendst/Tasmota#15669] [Q&A] [answered]
Hi there, 
I was told in another  thread to use ESP32 and then Berry language.
OK, so I  try to use ESP32, have 2 boards, WROOM32 (NodeMCU32) and ESP32-CAM (+carrier boardd with serial to USB).
Now I tried to flash Tasmota to boards:
- Node with Tasmota32 - -PL.bin
- Camera with Tasmota32-cam.bin,
using esp-idf as  od 06.2020.
Exact command line is:
./esptool.py --chip esp32 --before default_reset --after hard_reset --port /dev/ttyUSB5 --baud 921600 write_flash -z --flash_mode dout 0x0 /opt/esp-open-sdk.old/esptool/tasmota32-PL_11.1.0.bin
Resulted OK: 
> [laurent@laptop4 esptool]$ pwd
> /opt/ESP.disk/esp-idf/components/esptool_py/esptool
> [laurent@laptop4 esptool]$ ./esptool.py --chip esp32 --before default_reset --after hard_reset --port /dev/ttyUSB5 --baud 921600 write_flash -z --flash_mode dout 0x0 /opt/esp-open-sdk.old/esptool/tasmota32-PL_11.1.0.bin
> esptool.py v2.3.1
> Connecting........____
> Chip is ESP32D0WDQ6 (revision (unknown 0xa))
> Features: WiFi, BT, Dual Core, VRef calibration in efuse
> Uploading stub...
> Running stub...
> Stub running...
> Changing baud rate to 921600
> Changed.
> Configuring flash size...
> Auto-detected Flash size: 4MB
> Compressed 1429664 bytes to 1001018...
> Wrote 1429664 bytes (1001018 compressed) at 0x00000000 in 13.1 seconds (effective 874.7 kbit/s)...
> Hash of data verified.
> 
> Leaving...
> Hard resetting via RTS pin...
> [laurent@laptop4 esptool]$ 

But Minicom show garbage, and no WiFi activity:

> rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 
ets Jun  8 2016 00:22:57
<repeated>

What am I missing?

**Comment by barbudor** [comment, marked as answer]
> What am I missing?

To read https://tasmota.github.io/docs/ESP32/#flashing

Use the web flasher. It takes care of all details

**Reply by laurentppol** [reply]
But I already have been flashig ESP32 with my own toolllset!
Only thing I remember, there was 3 regions of flash to write.
Have read what You propose.

**Reply by laurentppol** [reply]
> Your browser does not support Web Serial.
Open this page in Google Chrome or
Microsoft Edge instead (but not on your iOS device).

**Reply by barbudor** [reply]
If you want to flash with esptool do as stated on the doc page and flash tasmota32.factory.bin
You can upgrade to -PL afterwards

I wonder how you were trying to flash to get such message...

**Reply by barbudor** [reply]
There's even a tasmota32-PL.factory.bin on the release page https://github.com/arendst/Tasmota/releases/tag/v11.1.0

**Reply by sfromis** [reply]
To flash ESP32 without using the `.factory` images, you'd need a more complex operation with 3 additional files at the right locations. The `.factory` images makes this much simpler, and very similar to flashing ESP8266.

**Reply by laurentppol** [reply]
OK, it worked at least for NodeMCU (will test webcam later), 
but this leaves at least 2 questions open:
- POOR documentation,
- WHY it  is hidden to users, that they have at least 3 partions in flash?

- boot, - partition table, and only then user / data/ SPIFFS partition.

**Reply by laurentppol** [reply]
> To flash ESP32 without using the `.factory` images, you'd need a more complex operation with 3 additional files at the right locations. The `.factory` images makes this much simpler, and very similar to flashing ESP8266.

Where is this HIDDEN?

**Reply by s-hadinger** [reply]
Documentation is open source. If you think it is poor, please submit pull requests to improve it. 

Nothing is hidden. It's all there. 

As mentioned the web flasher is very easy to use and handles all this automatically. This is why we propose it at first and try to not show all the complexity behind.

**Reply by sfromis** [reply]
Docs is very clear about how you're supposed to flash ESP32 chips if you prefer `esptool.py`, using the `.factory` image. Do not expect every non-recommended procedure to to be included to complicate the docs without benefit. I merely mentioned a bit of the backstory from before the `.factory` images were created, where it was more complex to do flashing of ESP32. By now, I'd call this setup "advanced usage", and outside the scope of basic guidance about recommended way of flashing.

Nothing hidden, but not every detail about how to handle ESP32 chips is or even should be included in Tasmota docs, only what's recommended by the project. You can go on from there, if you like, but then you're on your own.

**Reply by laurentppol** [reply]
@sfromis You are almost right, there IS information but somewheat hidden:
if You go the path I find most intuitive - "docs" -> "ESP32" -> "flashing", You get here: https://tasmota.github.io/docs/ESP32/#flashing, and there IS information, "flash tasmota32.FACTORY.bin" but there is NO information, You HAVE TO use "FACTORY"  image for the first time! (1), and more, there is NO information WHERE to look for this image, as they are NOT in "standard location" of precompiled images - http://ota.tasmota.com/tasmota32/release/ (just below warning about power supply). For FACTORY files You have to go to "precompiled", then to github page and then scroll down to "assets"! (2)
Sorry, don't find it intuitive! But OK, this is only  when You choose to use esptool.py.

**Reply by sfromis** [reply]
Now, the `.factory` images are included on http://ota.tasmota.com/tasmota32/ in addition to https://github.com/tasmota/install/tree/main/firmware/development and so on.
